### PR TITLE
[multi-test] Add staking and distribution module

### PR DESCRIPTION
**Pull request by ueco-jb**
Closes https://github.com/CosmWasm/cw-plus/issues/753

Implementation of Staking and Distribution modules in multitest package, which allows now to test staking and delegation.

### CODE REVIEW

File: packages/multi-test/src/staking.rs [hunk 1]
```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
...
+        msg: StakingMsg,
+    ) -> AnyResult<AppResponse> {
+        let mut staking_storage = prefixed(storage, NAMESPACE_STAKING);
+        match msg {
+            StakingMsg::Delegate { validator, amount } => {
+                let events = vec![Event::new("delegate")
+                    .add_attribute("recipient", &validator)
+                    .add_attribute("sender", &sender)
+                    .add_attribute("amount", format!("{}{}", amount.amount, amount.denom))];
+                self.add_stake(&mut staking_storage, sender, vec![amount])?;
```
Comment (ethanfrey) [inline packages/multi-test/src/staking.rs:126]: TODO: we need to deduct from the account that was validating (from the bank module).
I am pretty sure this is possible, I just need some time to do this (and then I will push it)

File: packages/multi-test/src/staking.rs [hunk 2]
```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
...
+                Ok(AppResponse { events, data: None })
+            }
+            StakingMsg::Undelegate { validator, amount } => {
+                let events = vec![Event::new("undelegate")
+                    .add_attribute("from", &validator)
+                    .add_attribute("to", &sender)
+                    .add_attribute("amount", format!("{}{}", amount.amount, amount.denom))];
+                self.remove_stake(&mut staking_storage, sender, vec![amount])?;
```
Comment (ethanfrey) [inline packages/multi-test/src/staking.rs:134]: TODO: need to do the bank send back to this account as above

File: packages/multi-test/src/staking.rs [hunk 3]
```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
...
+                // for now validator is ignored, as I want to support only one validator
+                let delegator = api.addr_validate(&delegator)?;
+                let stakes = match self.get_stakes(&staking_storage, &delegator) {
+                    Ok(stakes) => stakes[0].clone(),
+                    Err(_) => {
+                        let response = DelegationResponse { delegation: None };
+                        return Ok(to_binary(&response)?);
+                    }
+                };
+                // set fixed reward ratio 1:10 per delegated amoutn
```
Comment (ethanfrey) [inline packages/multi-test/src/staking.rs:175]: This should definitely be configurable somehow (via sudo or via custom "setup" calls).
Basically, nothing is free, you set it up defining how much rewards, then contract withdraws.
Can simulate contract withdrawing twice in one block, with 1 day between, with.1 year between that way

File: packages/multi-test/src/staking.rs [hunk 4]
```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
...
+        _block: &BlockInfo,
+        sender: Addr,
+        msg: DistributionMsg,
+    ) -> AnyResult<AppResponse> {
+        let staking_storage = prefixed(storage, NAMESPACE_STAKING);
+        match msg {
+            // For now it ignores validator as I want to support only one
+            DistributionMsg::WithdrawDelegatorReward { validator } => {
+                let stakes = self.get_stakes(&staking_storage, &sender)?[0].clone();
+                // set fixed reward ratio 1:10 per delegated amoutn
```
Comment (ethanfrey) [inline packages/multi-test/src/staking.rs:233]: As above in the staking query, need a good way to configure how many rewards.
Maybe we do it in this module and the staking module calls this internally to service this request?

**Review by ethanfrey — COMMENTED** [review]
Nice start.
The two big TODO are obvious missing points you mentioned to me.
I designed this to be possible. Probably need 30 minutes in the morning to push such a commit that will compile and do that

**Comment by chipshort** [comment]
This now allows adding multiple validators and the rewards are calculated based on a configurable APR, the validator's commission and the elapsed time, so you can just change `block.time` and send the withdrawal message. I also implemented the unbonding period (tough only when undelegating, not when redelegating). To trigger the actual payout when unbonding, a sudo message has to be sent to process the unbonding queue.
Not sure if this is a good solution, but I couldn't come up with a better way. 

Tarpaulin seems to be a bit silly. It marks a lot of covered lines as uncovered.

**Comment by JakeHartnell** [comment]
Can't wait for this!

**Review by uint — APPROVED** [review]
The code looks fine! I can't easily verify this is precisely how staking/distribution works, so I'm taking that on faith.

**Comment by ethanfrey** [comment]
@ueco-jb can you add "Closes #753 " to the description, and maybe update it (I think it isn't WIP anymore)

**Comment by ueco-jb** [comment]
@chipshort In reward formula remember to subtract the validator commision
`reward * (1 - commision)`

**Comment by chipshort** [comment]
> @chipshort In reward formula remember to subtract the validator commision `reward * (1 - commision)`

That should already be the case in `calculate_rewards`, albeit I calculated it in a slightly different way:
```rust
let commission = reward * validator_commission;

reward - commission
```

**Pull request merged**